sdk/log/batch_test.go

```diff
@@ -4,15 +4,169 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	b := NewBatchingProcessor(slowExporter{})
+	err := b.ForceFlush(ctx)
+	require.ErrorIs(t, err, context.Canceled)
+}
+
+func TestBatchingProcessorOnEmit(t *testing.T) {
+	e := new(exporter)
+	b := NewBatchingProcessor(e)
+	t.Cleanup(func() {
+		assert.NoError(t, b.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = b.OnEmit(context.Background(), r)
+	require.False(t, e.exportCalled, "exporter Export called")
+}
+
+func TestBatchingProcessorOnEmitForceFlush(t *testing.T) {
+	e := new(exporter)
+	b := NewBatchingProcessor(e)
+	t.Cleanup(func() {
+		assert.NoError(t, b.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = b.OnEmit(context.Background(), r)
+	_ = b.ForceFlush(context.Background())
+	require.True(t, e.exportCalled, "exporter Export not called")
+	assert.Equal(t, []Record{r}, e.records)
+}
+
+func TestBatchingProcessorOnEmitTick(t *testing.T) {
+	e := new(syncExporter)
+	s := NewBatchingProcessor(e, WithExportInterval(time.Millisecond))
+	t.Cleanup(func() {
+		assert.NoError(t, s.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = s.OnEmit(context.Background(), r)
+
+	require.Eventually(t, func() bool { return e.exportCalled.Load() }, time.Second, time.Millisecond, "exporter Export not called")
+	assert.Equal(t, []Record{r}, e.Records())
+}
+
+func TestBatchingProcessorOnFullQueue(t *testing.T) {
+	e := new(exporter)
+	b := NewBatchingProcessor(e, WithMaxQueueSize(1))
+	t.Cleanup(func() {
+		assert.NoError(t, b.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = b.OnEmit(context.Background(), r)
+	var r2 Record
+	r2.SetSeverityText("dropped")
+	_ = b.OnEmit(context.Background(), r2)
+
+	_ = b.ForceFlush(context.Background())
+	assert.Equal(t, []Record{r}, e.records)
+}
+
+func TestBatchingProcessorOnFullBatch(t *testing.T) {
+	e := new(exporter)
+	b := NewBatchingProcessor(e, WithExportMaxBatchSize(1))
+	t.Cleanup(func() {
+		assert.NoError(t, b.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = b.OnEmit(context.Background(), r)
+	var r2 Record
+	r2.SetSeverityText("on next export")
+	_ = b.OnEmit(context.Background(), r2)
+
+	_ = b.ForceFlush(context.Background())
+	assert.Equal(t, []Record{r}, e.records)
+
+	_ = b.ForceFlush(context.Background())
+	assert.Equal(t, []Record{r, r2}, e.records)
+}
+
+func TestBatchingProcessorConcurrentSafe(t *testing.T) {
+	const goRoutineN = 10
+
+	var wg sync.WaitGroup
+	wg.Add(goRoutineN)
+
+	var r Record
+	r.SetSeverityText("test")
+	ctx := context.Background()
+	s := NewBatchingProcessor(nil)
+	t.Cleanup(func() {
+		assert.NoError(t, s.Shutdown(context.Background()))
+	})
+	for i := 0; i < goRoutineN; i++ {
+		go func() {
+			defer wg.Done()
+
+			_ = s.OnEmit(ctx, r)
+			_ = s.Enabled(ctx, r)
+			_ = s.Shutdown(ctx)
```

Comment:
This may be called before all the other goroutine have a chance to run, meaning the concurrency check is going to be for a shutdown processor.